addons/stock/models/stock_quant.py

```diff
@@ -1309,6 +1309,111 @@ def _get_quants_action(self, domain=None, extend=False):
+                return ''
+            tracking_ai = '21' if self.tracking == 'serial' else '10'
+            barcode += tracking_ai + self.lot_id.name
+        return barcode
+
+    def get_aggregated_barcodes(self):
+        """ Generates and aggregates quants' barcodes. Depending of the number of quants and the
+        length of their barcode encodings, there will be one or more aggregated barcodes (returned
+        as a list). This method uses the config parameters `stock.agg_barcode_max_length` to know
+        the limit of a single aggregated barcode (400 by default) and `stock.barcode_separator` to
+        know which character to use to separate individual encodings.
+        Without that last parameter, this method can't work and will return a empty list.
+
+        :return: list
+        """
+        agg_barcode_max_length = int(self.env['ir.config_parameter'].sudo().get_param('stock.agg_barcode_max_length', 400))
+        barcode_separator = self.env['ir.config_parameter'].sudo().get_param('stock.barcode_separator')
+        if not barcode_separator:
+            return []  # A barcode separator is mandatory to be able to aggregate barcodes.
+
+        eol_char = '\t'  # Added at the end of aggregated barcodes to end `barcode_scanned` event.
+        composite_barcodes = []
+        composite_barcode = ""
+
+        # Searches all GS1 rules linked to an UoM other than Unit and retrieves their AI.
+        uom_unit_id = self.env.ref('uom.product_uom_unit').id
+        gs1_quantity_rules = self.env['barcode.rule'].search([
+            ('associated_uom_id', '!=', False),
+            ('associated_uom_id', '!=', uom_unit_id),
+            ('is_gs1_nomenclature', '=', True)]
+        )
+        gs1_quantity_rules_ai_by_uom = {}
+
+        for rule in gs1_quantity_rules:
+            decimal = str(len(f'{rule.associated_uom_id.rounding:.10f}'.rstrip('0').split('.')[1]))
+            rule_ai = rule.pattern[1:4] + decimal
+            gs1_quantity_rules_ai_by_uom[rule.associated_uom_id.id] = rule_ai
+
+        previous_product = self.env['product.product']
+        for quant in self:
+            if not quant.product_id.barcode:
+                continue
+            barcode = ""
+            # In case the quant product's barcode is not GS1 compliant, add it first,
+            # so that the lots and qty barcodes that follow it will be used for this product.
+            if previous_product != quant.product_id:
+                previous_product = quant.product_id
+                if not quant.product_id.valid_ean:
+                    barcode += quant.product_id.barcode + barcode_separator
+            # Gets quant's barcode (either a GS1 barcode or only a serial number.)
+            quant_gs1_barcode = quant._get_gs1_barcode(gs1_quantity_rules_ai_by_uom)
+            if quant_gs1_barcode:
+                barcode += quant_gs1_barcode
+            elif quant.tracking == 'serial':
+                barcode += quant.lot_id.name
```

Comment:
About the code suggestion, I did that instead because we don't want to add the separator if there is no barcode already defined
```python
                if not quant.product_id.valid_ean:
                    barcode += quant.product_id.barcode
            # Gets quant's barcode (either a GS1 barcode or only a serial number.)
            quant_gs1_barcode = quant._get_gs1_barcode(gs1_quantity_rules_ai_by_uom)
            if quant_gs1_barcode:
                barcode += (barcode_separator if barcode else '') + quant_gs1_barcode
            elif quant.tracking == 'serial':
                barcode += (barcode_separator if barcode else '') + quant.lot_id.name
```

> Also, does it make sense that non-tracked products without a valid ean don't have their quantity GS1 barcode added? It seems more confusing to me that a QR Code of just their product barcode is printed

Yes, I did that otherwise the qty will be wrong when scanned.
I thought of doing that:
- First, we add the product barcode followed by the separator;
- Then, we add the qty as a GS1 barcode.

The issue with that is, when the aggregate barcode will be scanned, that will happen:
- The aggregate barcode will be split thanks to the separator and we will have to separate barcodes:
  1. The product barcode
  2. The quantity barcode
- The first barcode will be processed, a line for the product will be added with 1 qty;
- The second barcode will be processed, adding the qty to the selected line -> Since the line has already 1 qty, let say the quant is for 25x product, we will end with 26 qty (1 + 25)
This issue doesn't happen with tracked product because the qty of tracked product is incremented when the lot/serial is scanned, not the product.

Reply:
Fair enough for both. I'm not sure if it's better to write an code comment explanation and/or in the commit message that it's expected that non-GS1 compliant product barcodes won't have a qty printed with them :thinking: I predict that users will be confused about it and tech support people will have to repeatedly explain to them why